src/libs/actions/PolicyConnectiont.ts

```diff
@@ -0,0 +1,83 @@
+import Onyx from 'react-native-onyx';
+import type {OnyxUpdate} from 'react-native-onyx';
+import * as API from '@libs/API';
+import {WRITE_COMMANDS} from '@libs/API/types';
+import * as ErrorUtils from '@libs/ErrorUtils';
+import CONST from '@src/CONST';
+import ONYXKEYS from '@src/ONYXKEYS';
+import type {QBOConnectionConfig} from '@src/types/onyx/Policy';
+
+function updatePolicyConnectionConfig({
+    policyID,
+    connectionName,
+    updatedField,
+    updatedPartialConfig,
+    originalPartialConfig,
+}: {
+    policyID: string;
+    connectionName: string;
+    updatedField: string;
+    updatedPartialConfig: Partial<QBOConnectionConfig>;
+    originalPartialConfig: Partial<QBOConnectionConfig>;
+}) {
+    const optimisticData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                pendingFields: {
+                    [updatedField]: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
+                },
+                connections: {
+                    [connectionName]: {
+                        config: updatedPartialConfig,
+                    },
+                },
+            },
+        },
+    ];
+
+    const successData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                connections: {
+                    [connectionName]: {
+                        config: updatedPartialConfig,
+                    },
+                },
+            },
+        },
+    ];
+
+    const failureData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                errorFields: {
+                    [updatedField]: ErrorUtils.getMicroSecondOnyxError('workspace.connection.error.genericUpdate'),
+                },
+                connections: {
+                    [connectionName]: {
+                        config: originalPartialConfig,
+                    },
```

Comment:
@aldo-expensify I'm struggling with how to set the `failureData`.

We decided to send the partial connections object with the changed properties.

In the case of the failed operation, I don't know how we can bring the connections object back to the original state. Do you have an idea of how to go about this?

Reply:
Assuming that:
- you have the full connection data where you call `updatePolicyConnectionConfig`
- you constructed `updatedPartialConfig` also where you called `updatePolicyConnectionConfig`

I think you should be able to construct `originalPartialConfig` the same way you constructed `updatedPartialConfig`, but instead of using a value from the input, you use the value you find inside the connection data. Does that make sense?

Reply:
Unrelated: there is a typo in the name src/libs/actions/PolicyConnectiont.ts

Reply:
> I think you should be able to construct originalPartialConfig the same way you constructed updatedPartialConfig, but instead of using a value from the input, you use the value you find inside the connection data. Does that make sense?

This was the first solution that came to my mind. However, if there are other updates operations that succeed, the failure data ends up overwriting changes made in the successful updates. I don't know how to account for this case 🤔

Reply:
ah now I see what you mean. I can put the partial original connections object that corresponds to the changed part